### PR TITLE
Refactors how movetypes are added and removed, No timers this time.

### PR DESCRIPTION
**Pull request by Ghommie**
## About The Pull Request
Pretty much #54963 but without the abhominable amount of timers that lagged live servers so much.
EDIT: Made this an element upon Jared's request.

## Why It's Good For The Game
A way to avoid conflict from multiple sources of movement types.
Tested it again multiple times.

## Changelog
:cl:
fix: jitteriness, melee attack animations and resting/standing up should no longer momentarily disable the floating movetype.
/:cl:

### CODE REVIEW

File: code/__DEFINES/traits.dm [hunk 1]
```diff
@@ -219,6 +219,35 @@ Remember to update _globalvars/traits.dm if you're adding/removing/renaming trai
...
+		AM.RegisterSignal(AM, GLOB.movement_type_trait_add_signals, /atom/movable/.proc/on_movement_type_trait_gain);\
+		AM.RegisterSignal(AM, GLOB.movement_type_trait_remove_signals, /atom/movable/.proc/on_movement_type_trait_loss);\
+		AM.RegisterSignal(AM, SIGNAL_ADDTRAIT(TRAIT_NO_FLOATING_ANIM), /atom/movable/.proc/on_no_floating_anim_trait_gain, TRUE);\
+		AM.RegisterSignal(AM, SIGNAL_REMOVETRAIT(TRAIT_NO_FLOATING_ANIM), /atom/movable/.proc/on_no_floating_anim_trait_loss, TRUE);\
+		AM.has_movement_type_signals = TRUE;\
+		if(HAS_TRAIT(AM, TRAIT_NO_FLOATING_ANIM)){\
+			AM.floating_anim_status = NEVER_FLOATING_ANIM\
+		};\
+	};\
+	ADD_TRAIT(AM, trait, source)
```
Comment (Jared-Fogle) [inline code/__DEFINES/traits.dm:240]: This kind of stuff is concerning to me. I was thinking something more in line with a floating element (which would also handle the processing in one bulk there).

Comment (Ghommie) [reply code/__DEFINES/traits.dm:240]: Umh, I'll see if I can make that an element. Good catch.

File: code/datums/elements/movetype_handler.dm [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+	var/list/paused_floating_anim_atoms = list()
+
+/datum/element/movetype_handler/Attach(datum/target)
+	. = ..()
+	if(!ismovable(target))
+		return ELEMENT_INCOMPATIBLE
+	if(attached_atoms[target]) //Already attached.
+		return
+
+	var/atom/movable/AM = target
```
Comment (Jared-Fogle) [inline code/datums/elements/movetype_handler.dm:20]: Better variable names.

File: code/_globalvars/traits.dm [hunk 3]
```diff
@@ -180,3 +187,19 @@ GLOBAL_LIST(trait_name_map)
...
+	TRAIT_MOVE_PHASING = PHASING
+	))
+
+GLOBAL_LIST_INIT(movement_type_trait_add_signals, set_movement_type_trait_signals(SIGNAL_ADDTRAIT("")))
+GLOBAL_LIST_INIT(movement_type_trait_remove_signals, set_movement_type_trait_signals(SIGNAL_REMOVETRAIT("")))
+
+/proc/set_movement_type_trait_signals(signal_prefix)
+	. = list()
+	for(var/trait in GLOB.movement_type_trait_to_flag)
+		. += "[signal_prefix][trait]"
```
Comment (Jared-Fogle) [inline code/_globalvars/traits.dm:205]: This seems like a flimsy exploitation of how SIGNAL_ADDTRAIT/SIGNAL_REMOVETRAIT work internally--there's no guarantee that it'll be in this format forever. 

I think you should change the movement_type_trait_add_signals/movement_type_trait_remove_signals to be lazy lists, and have a wrapper to get them (like `get_movement_type_trait_add_signals`) that will dynamically set those lists and use `SIGNAL_ADDTRAIT(...)` and `SIGNAL_REMOVETRAIT(...)` how they were intended to be used.

File: code/datums/elements/movetype_handler.dm [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+	SIGNAL_HANDLER
+	if(paused_floating_anim_atoms[source] < world.time + timer)
+		stop_floating(source)
+		if(!length(paused_floating_anim_atoms))
+			START_PROCESSING(SSdcs, src) //1 second tickrate.
+		paused_floating_anim_atoms[source] = world.time + timer
+
+/datum/element/movetype_handler/process()
+	for(var/a in paused_floating_anim_atoms)
+		var/atom/movable/AM = a
```
Comment (Jared-Fogle) [inline code/datums/elements/movetype_handler.dm:88]: Better variable names. Something like this is good:

```
	for(var/_floating in paused_floating_anim_atoms)
		var/atom/movable/floating = _floating
```

Comment (Jared-Fogle) [reply code/datums/elements/movetype_handler.dm:88]: I don't want to comment on all of them specifically, but just change every variable you create that's `AM`, `L`, `M`, etc too be more exact to what they actually are.

File: code/game/objects/buckling.dm [hunk 5]
```diff
@@ -146,10 +150,24 @@
...
 	buckled_mobs -= buckled_mob
+	if(anchored)
+		REMOVE_TRAIT(buckled_mob, TRAIT_NO_FLOATING_ANIM, BUCKLED_TRAIT)
+	if(!length(buckled_mobs))
+		UnregisterSignal(src, COMSIG_MOVABLE_SET_ANCHORED, .proc/on_set_anchored)
 	SEND_SIGNAL(src, COMSIG_MOVABLE_UNBUCKLE, buckled_mob, force)
 
 	post_unbuckle_mob(.)
 
+/atom/movable/proc/on_set_anchored(atom/movable/source, anchorvalue)
```
Comment (Jared-Fogle) [inline code/game/objects/buckling.dm:161]: Needs `SIGNAL_HANDLER`.

File: code/game/objects/buckling.dm [hunk 6]
```diff
@@ -146,10 +150,24 @@
...
+	for(var/M in buckled_mobs)
+		if(!M)
+			continue
+		var/mob/living/buckled_mob = M
```
Comment (Jared-Fogle) [inline code/game/objects/buckling.dm:165]:
```suggestion
	for(var/_buckled_mob in buckled_mobs)
		if(!_buckled_mob)
			continue
		var/mob/living/buckled_mob = _buckled_mob
```

File: code/modules/mob/living/init_signals.dm [hunk 7]
```diff
@@ -180,3 +182,13 @@
 		throw_alert("succumb", /atom/movable/screen/alert/succumb)
 	else
 		clear_alert("succumb")
+
+///From [element/movetype_handler/on_movement_type_trait_gain()]
+/mob/living/proc/on_movement_type_flag_enabled(datum/source, trait)
+	. = ..()
```
Comment (Jared-Fogle) [inline code/modules/mob/living/init_signals.dm:188]: Read the linters.

**Review by Mothblocks — CHANGES_REQUESTED** [review]

**Comment by Jared-Fogle** [comment]
TM'd on Bagil.

**Comment by optimumtact** [comment]
well?

**Comment by Jared-Fogle** [comment]
It works.

**Pull request merged**